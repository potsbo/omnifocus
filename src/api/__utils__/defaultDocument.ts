const projects = () => {
  return [...Array(10).keys()].map((i) => {
    const projecId = `project-id-${i}`;
    const projectName = `project-name-${i}`;
    return {
      properties: () => ({ pcls: "Project" }),
      name: () => projectName,
      id: () => projecId,
      rootTask: () => {
        return {
          name: () => `rootTask-for-${projectName}`,
          id: () => `rootTask-for-${projecId}`,
          effectiveDueDate: () => null,
          completed: () => false,
          effectivelyCompleted: () => false,
          containingProject: () => {
            return {
              id: () => projecId,
              name: () => projectName,
            };
          },
          flagged: () => false,
          effectiveDeferDate: () => null,
        };
      },
    };
  });
};

projects.byId = (pid: string) => {
  return {
    properties: () => ({
      pcls: "Project",
    }),
    rootTask: () => {
      return {
        tasks: () => [
          {
            properties: () => ({ pcls: "Task" }),
            name: () => "foo",
            id: () => "bar",
            effectiveDueDate: () => null,
            completed: () => false,
            effectivelyCompleted: () => false,
            containingProject: () => {
              return {
                id: () => pid,
                name: () => "projectName",
                properties: () => ({
                  pcls: "Project",
                }),
              };
            },
            flagged: () => false,
            effectiveDeferDate: () => null,
          },
        ],
      };
    },
  };
};

export const defaultDocument = {
  projects,
  inboxTasks: [],
};
