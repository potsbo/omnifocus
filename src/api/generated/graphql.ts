import { GraphQLResolveInfo } from 'graphql';
import { GraphQLClient } from 'graphql-request';
import * as Dom from 'graphql-request/dist/types.dom';
import gql from 'graphql-tag';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type RequireFields<T, K extends keyof T> = Omit<T, K> & { [P in K]-?: NonNullable<T[P]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Project = {
  __typename?: 'Project';
  id: Scalars['String'];
  name: Scalars['String'];
  rootTask: Task;
};

export type Query = {
  __typename?: 'Query';
  flattenedTasks: Array<Task>;
  inboxTasks: Array<Task>;
  project?: Maybe<Project>;
};


export type QueryFlattenedTasksArgs = {
  available?: InputMaybe<Scalars['Boolean']>;
  flagged?: InputMaybe<Scalars['Boolean']>;
  limit?: InputMaybe<Scalars['Int']>;
  withEffectiveDueDate?: InputMaybe<Scalars['Boolean']>;
};


export type QueryInboxTasksArgs = {
  available?: InputMaybe<Scalars['Boolean']>;
  flagged?: InputMaybe<Scalars['Boolean']>;
};


export type QueryProjectArgs = {
  id: Scalars['String'];
};

export type Task = {
  __typename?: 'Task';
  completed: Scalars['Boolean'];
  containingProject?: Maybe<Project>;
  effectiveDueDate?: Maybe<Scalars['String']>;
  effectivelyCompleted: Scalars['Boolean'];
  flagged: Scalars['Boolean'];
  id: Scalars['String'];
  name: Scalars['String'];
  tasks?: Maybe<Array<Task>>;
};

export type TaskViewModelFragment = { __typename?: 'Task', name: string, id: string, effectiveDueDate?: string | null, completed: boolean, effectivelyCompleted: boolean, flagged: boolean, containingProject?: { __typename?: 'Project', id: string, name: string } | null };

export type GetTasksQueryVariables = Exact<{
  flagged?: InputMaybe<Scalars['Boolean']>;
  available?: InputMaybe<Scalars['Boolean']>;
  withEffectiveDueDate?: InputMaybe<Scalars['Boolean']>;
}>;


export type GetTasksQuery = { __typename?: 'Query', flattenedTasks: Array<{ __typename?: 'Task', name: string, id: string, effectiveDueDate?: string | null, completed: boolean, effectivelyCompleted: boolean, flagged: boolean, containingProject?: { __typename?: 'Project', id: string, name: string } | null }> };

export type GetInboxTasksQueryVariables = Exact<{
  flagged?: InputMaybe<Scalars['Boolean']>;
  available?: InputMaybe<Scalars['Boolean']>;
}>;


export type GetInboxTasksQuery = { __typename?: 'Query', inboxTasks: Array<{ __typename?: 'Task', name: string, id: string, effectiveDueDate?: string | null, completed: boolean, effectivelyCompleted: boolean, flagged: boolean, containingProject?: { __typename?: 'Project', id: string, name: string } | null }> };

export type GetTasksInProjectQueryVariables = Exact<{
  projectId: Scalars['String'];
}>;


export type GetTasksInProjectQuery = { __typename?: 'Query', project?: { __typename?: 'Project', rootTask: { __typename?: 'Task', tasks?: Array<{ __typename?: 'Task', name: string, id: string, effectiveDueDate?: string | null, completed: boolean, effectivelyCompleted: boolean, flagged: boolean, containingProject?: { __typename?: 'Project', id: string, name: string } | null }> | null } } | null };



export type ResolverTypeWrapper<T> = Promise<T> | T;


export type ResolverWithResolve<TResult, TParent, TContext, TArgs> = {
  resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> = ResolverFn<TResult, TParent, TContext, TArgs> | ResolverWithResolve<TResult, TParent, TContext, TArgs>;

export type ResolverFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => Promise<TResult> | TResult;

export type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => AsyncIterable<TResult> | Promise<AsyncIterable<TResult>>;

export type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<{ [key in TKey]: TResult }, TParent, TContext, TArgs>;
  resolve?: SubscriptionResolveFn<TResult, { [key in TKey]: TResult }, TContext, TArgs>;
}

export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
  resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}

export type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> =
  | SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs>
  | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;

export type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> =
  | ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>)
  | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;

export type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (
  parent: TParent,
  context: TContext,
  info: GraphQLResolveInfo
) => Maybe<TTypes> | Promise<Maybe<TTypes>>;

export type IsTypeOfResolverFn<T = {}, TContext = {}> = (obj: T, context: TContext, info: GraphQLResolveInfo) => boolean | Promise<boolean>;

export type NextResolverFn<T> = () => Promise<T>;

export type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (
  next: NextResolverFn<TResult>,
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = {
  Boolean: ResolverTypeWrapper<Scalars['Boolean']>;
  Int: ResolverTypeWrapper<Scalars['Int']>;
  Project: ResolverTypeWrapper<Project>;
  Query: ResolverTypeWrapper<{}>;
  String: ResolverTypeWrapper<Scalars['String']>;
  Task: ResolverTypeWrapper<Task>;
};

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = {
  Boolean: Scalars['Boolean'];
  Int: Scalars['Int'];
  Project: Project;
  Query: {};
  String: Scalars['String'];
  Task: Task;
};

export type ProjectResolvers<ContextType = any, ParentType extends ResolversParentTypes['Project'] = ResolversParentTypes['Project']> = {
  id?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  name?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  rootTask?: Resolver<ResolversTypes['Task'], ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type QueryResolvers<ContextType = any, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = {
  flattenedTasks?: Resolver<Array<ResolversTypes['Task']>, ParentType, ContextType, Partial<QueryFlattenedTasksArgs>>;
  inboxTasks?: Resolver<Array<ResolversTypes['Task']>, ParentType, ContextType, Partial<QueryInboxTasksArgs>>;
  project?: Resolver<Maybe<ResolversTypes['Project']>, ParentType, ContextType, RequireFields<QueryProjectArgs, 'id'>>;
};

export type TaskResolvers<ContextType = any, ParentType extends ResolversParentTypes['Task'] = ResolversParentTypes['Task']> = {
  completed?: Resolver<ResolversTypes['Boolean'], ParentType, ContextType>;
  containingProject?: Resolver<Maybe<ResolversTypes['Project']>, ParentType, ContextType>;
  effectiveDueDate?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  effectivelyCompleted?: Resolver<ResolversTypes['Boolean'], ParentType, ContextType>;
  flagged?: Resolver<ResolversTypes['Boolean'], ParentType, ContextType>;
  id?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  name?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  tasks?: Resolver<Maybe<Array<ResolversTypes['Task']>>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type Resolvers<ContextType = any> = {
  Project?: ProjectResolvers<ContextType>;
  Query?: QueryResolvers<ContextType>;
  Task?: TaskResolvers<ContextType>;
};


export const TaskViewModelFragmentDoc = gql`
    fragment TaskViewModel on Task {
  name
  id
  effectiveDueDate
  completed
  effectivelyCompleted
  containingProject {
    id
    name
  }
  flagged
}
    `;
export const GetTasksDocument = gql`
    query GetTasks($flagged: Boolean, $available: Boolean, $withEffectiveDueDate: Boolean) {
  flattenedTasks(
    flagged: $flagged
    available: $available
    withEffectiveDueDate: $withEffectiveDueDate
  ) {
    ...TaskViewModel
  }
}
    ${TaskViewModelFragmentDoc}`;
export const GetInboxTasksDocument = gql`
    query GetInboxTasks($flagged: Boolean, $available: Boolean) {
  inboxTasks(flagged: $flagged, available: $available) {
    ...TaskViewModel
  }
}
    ${TaskViewModelFragmentDoc}`;
export const GetTasksInProjectDocument = gql`
    query GetTasksInProject($projectId: String!) {
  project(id: $projectId) {
    rootTask {
      tasks {
        ...TaskViewModel
      }
    }
  }
}
    ${TaskViewModelFragmentDoc}`;

export type SdkFunctionWrapper = <T>(action: (requestHeaders?:Record<string, string>) => Promise<T>, operationName: string) => Promise<T>;


const defaultWrapper: SdkFunctionWrapper = (action, _operationName) => action();

export function getSdk(client: GraphQLClient, withWrapper: SdkFunctionWrapper = defaultWrapper) {
  return {
    GetTasks(variables?: GetTasksQueryVariables, requestHeaders?: Dom.RequestInit["headers"]): Promise<GetTasksQuery> {
      return withWrapper((wrappedRequestHeaders) => client.request<GetTasksQuery>(GetTasksDocument, variables, {...requestHeaders, ...wrappedRequestHeaders}), 'GetTasks');
    },
    GetInboxTasks(variables?: GetInboxTasksQueryVariables, requestHeaders?: Dom.RequestInit["headers"]): Promise<GetInboxTasksQuery> {
      return withWrapper((wrappedRequestHeaders) => client.request<GetInboxTasksQuery>(GetInboxTasksDocument, variables, {...requestHeaders, ...wrappedRequestHeaders}), 'GetInboxTasks');
    },
    GetTasksInProject(variables: GetTasksInProjectQueryVariables, requestHeaders?: Dom.RequestInit["headers"]): Promise<GetTasksInProjectQuery> {
      return withWrapper((wrappedRequestHeaders) => client.request<GetTasksInProjectQuery>(GetTasksInProjectDocument, variables, {...requestHeaders, ...wrappedRequestHeaders}), 'GetTasksInProject');
    }
  };
}
export type Sdk = ReturnType<typeof getSdk>;