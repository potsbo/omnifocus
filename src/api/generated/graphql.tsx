import { GraphQLResolveInfo } from "graphql";
import { RequestInit } from "graphql-request/dist/types.dom";
import { useQuery, UseQueryOptions } from "react-query";
import { fetch } from "../fetch";
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Project = {
  __typename?: "Project";
  id: Scalars["String"];
  name: Scalars["String"];
};

export type Query = {
  __typename?: "Query";
  flattenedTasks: Array<Maybe<Task>>;
};

export type QueryFlattenedTasksArgs = {
  available?: InputMaybe<Scalars["Boolean"]>;
  flagged?: InputMaybe<Scalars["Boolean"]>;
  limit?: InputMaybe<Scalars["Int"]>;
};

export type Task = {
  __typename?: "Task";
  completed: Scalars["Boolean"];
  containingProject?: Maybe<Project>;
  effectiveDueDate?: Maybe<Scalars["String"]>;
  effectivelyCompleted: Scalars["Boolean"];
  flagged: Scalars["Boolean"];
  id: Scalars["String"];
  name: Scalars["String"];
};

export type GetFlaggedTasksQueryVariables = Exact<{ [key: string]: never }>;

export type GetFlaggedTasksQuery = {
  __typename?: "Query";
  flattenedTasks: Array<{
    __typename?: "Task";
    name: string;
    id: string;
    effectiveDueDate?: string | null;
    completed: boolean;
    effectivelyCompleted: boolean;
    flagged: boolean;
    containingProject?: { __typename?: "Project"; id: string; name: string } | null;
  } | null>;
};

export type ResolverTypeWrapper<T> = Promise<T> | T;

export type ResolverWithResolve<TResult, TParent, TContext, TArgs> = {
  resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> =
  | ResolverFn<TResult, TParent, TContext, TArgs>
  | ResolverWithResolve<TResult, TParent, TContext, TArgs>;

export type ResolverFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => Promise<TResult> | TResult;

export type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => AsyncIterable<TResult> | Promise<AsyncIterable<TResult>>;

export type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<{ [key in TKey]: TResult }, TParent, TContext, TArgs>;
  resolve?: SubscriptionResolveFn<TResult, { [key in TKey]: TResult }, TContext, TArgs>;
}

export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
  resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}

export type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> =
  | SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs>
  | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;

export type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> =
  | ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>)
  | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;

export type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (
  parent: TParent,
  context: TContext,
  info: GraphQLResolveInfo
) => Maybe<TTypes> | Promise<Maybe<TTypes>>;

export type IsTypeOfResolverFn<T = {}, TContext = {}> = (
  obj: T,
  context: TContext,
  info: GraphQLResolveInfo
) => boolean | Promise<boolean>;

export type NextResolverFn<T> = () => Promise<T>;

export type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (
  next: NextResolverFn<TResult>,
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = {
  Boolean: ResolverTypeWrapper<Scalars["Boolean"]>;
  Int: ResolverTypeWrapper<Scalars["Int"]>;
  Project: ResolverTypeWrapper<Project>;
  Query: ResolverTypeWrapper<{}>;
  String: ResolverTypeWrapper<Scalars["String"]>;
  Task: ResolverTypeWrapper<Task>;
};

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = {
  Boolean: Scalars["Boolean"];
  Int: Scalars["Int"];
  Project: Project;
  Query: {};
  String: Scalars["String"];
  Task: Task;
};

export type ProjectResolvers<
  ContextType = any,
  ParentType extends ResolversParentTypes["Project"] = ResolversParentTypes["Project"]
> = {
  id?: Resolver<ResolversTypes["String"], ParentType, ContextType>;
  name?: Resolver<ResolversTypes["String"], ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type QueryResolvers<
  ContextType = any,
  ParentType extends ResolversParentTypes["Query"] = ResolversParentTypes["Query"]
> = {
  flattenedTasks?: Resolver<
    Array<Maybe<ResolversTypes["Task"]>>,
    ParentType,
    ContextType,
    Partial<QueryFlattenedTasksArgs>
  >;
};

export type TaskResolvers<
  ContextType = any,
  ParentType extends ResolversParentTypes["Task"] = ResolversParentTypes["Task"]
> = {
  completed?: Resolver<ResolversTypes["Boolean"], ParentType, ContextType>;
  containingProject?: Resolver<Maybe<ResolversTypes["Project"]>, ParentType, ContextType>;
  effectiveDueDate?: Resolver<Maybe<ResolversTypes["String"]>, ParentType, ContextType>;
  effectivelyCompleted?: Resolver<ResolversTypes["Boolean"], ParentType, ContextType>;
  flagged?: Resolver<ResolversTypes["Boolean"], ParentType, ContextType>;
  id?: Resolver<ResolversTypes["String"], ParentType, ContextType>;
  name?: Resolver<ResolversTypes["String"], ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type Resolvers<ContextType = any> = {
  Project?: ProjectResolvers<ContextType>;
  Query?: QueryResolvers<ContextType>;
  Task?: TaskResolvers<ContextType>;
};

export const GetFlaggedTasksDocument = `
    query GetFlaggedTasks {
  flattenedTasks(flagged: true) {
    name
    id
    effectiveDueDate
    completed
    effectivelyCompleted
    containingProject {
      id
      name
    }
    flagged
  }
}
    `;
export const useGetFlaggedTasksQuery = <TData = GetFlaggedTasksQuery, TError = unknown>(
  variables?: GetFlaggedTasksQueryVariables,
  options?: UseQueryOptions<GetFlaggedTasksQuery, TError, TData>
) =>
  useQuery<GetFlaggedTasksQuery, TError, TData>(
    variables === undefined ? ["GetFlaggedTasks"] : ["GetFlaggedTasks", variables],
    fetch<GetFlaggedTasksQuery, GetFlaggedTasksQueryVariables>(GetFlaggedTasksDocument, variables),
    options
  );
