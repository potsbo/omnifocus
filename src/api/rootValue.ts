import { GraphQLResolveInfo } from "graphql";
import { genQuery } from "./query";

export const buildRootValue = (appName: string, runner: (code: string) => unknown) => {
  return new Proxy(
    {},
    {
      get: function (_, name: string) {
        // query
        if (name === "application") {
          return (_: unknown, _2: unknown, info: GraphQLResolveInfo) => {
            const q = genQuery(appName, info);
            return runner(q);
          };
        }

        // mutation
        return (_: unknown, _2: unknown, info: GraphQLResolveInfo) => {
          const q = genQuery(appName, info, "obj");
          const code = `
              const app = Application("OmniFocus");
              const obj = app.InboxTask({ name: "something" });
              app.defaultDocument().inboxTasks.push(obj);
              ${q}
          `;
          return runner(code);
        };
      },
    }
  );
};
