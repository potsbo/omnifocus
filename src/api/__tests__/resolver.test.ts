import { graphql, print } from "graphql";
import { defaultDocument } from "../__utils__/defaultDocument";
import { GetTasksInProjectDocument } from "../generated/graphql";
import gql from "graphql-tag";
import { loadSchemaSync } from "@graphql-tools/load";
import { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
import { join } from "path";
import { buildRootValue } from "../rootValue";

export const schema = loadSchemaSync(join(__dirname, "..", "..", "..", "assets", "schema.graphql"), {
  loaders: [new GraphQLFileLoader()],
});

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fail(reason: any) {
  throw new Error(reason);
}

const rootValue = buildRootValue("OmniFocus", (query) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const Application = (_: string) => ({
    defaultDocument: () => defaultDocument,
    InboxTask: (props: Record<string, unknown>) => {
      return Object.fromEntries(Object.entries(props).map(([k, v]) => [k, () => v]));
    },
  });
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const Automation = {
    getDisplayString: (obj: { getDisplayString: () => string }) => {
      console.log(obj.getDisplayString());
      return obj.getDisplayString();
    },
  };

  return new Promise((resolve) => resolve(JSON.parse(eval(query)).result));
});

test("Resolve Connection Related Query", async () => {
  const document = gql`
    query Hoge {
      application {
        defaultDocument {
          projects {
            pageInfo {
              hasNextPage
            }
            edges {
              cursor
              node {
                name
                rootTask {
                  name
                  id
                  containingProject {
                    name
                    id
                  }
                }
              }
            }
          }
        }
      }
    }
  `;
  const { data, errors } = await graphql({
    schema,
    source: print(document),
    rootValue,
  });

  if (errors !== null && errors !== undefined) {
    fail(errors);
  }
  expect(data).toMatchSnapshot();
});

test("run for GetTasksInProjectDocument", async () => {
  const document = GetTasksInProjectDocument;
  const { data, errors } = await graphql({
    schema,
    source: print(document),
    rootValue,
    variableValues: { projectId: "some-project-id" },
  });

  if (errors !== null && errors !== undefined) {
    fail(errors);
  }

  expect(data).toMatchSnapshot();
});

test("simple mutation", async () => {
  const document = gql`
    mutation {
      pushInboxTask(name: "SomeTask") {
        name
      }
    }
  `;
  const { data, errors } = await graphql({
    schema,
    source: print(document),
    rootValue,
  });

  if (errors !== null && errors !== undefined) {
    fail(errors);
  }
  expect(data).toMatchSnapshot();
});
