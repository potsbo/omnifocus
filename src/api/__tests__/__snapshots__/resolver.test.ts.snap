// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolve Connection Related Query 1`] = `
Object {
  "defaultDocument": Object {
    "projects": Object {
      "edges": Array [
        Object {
          "cursor": "project-id-0",
          "node": Object {
            "name": "project-name-0",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-0",
                "name": "project-name-0",
              },
              "id": "rootTask-for-project-id-0",
              "name": "rootTask-for-project-name-0",
            },
          },
        },
        Object {
          "cursor": "project-id-1",
          "node": Object {
            "name": "project-name-1",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-1",
                "name": "project-name-1",
              },
              "id": "rootTask-for-project-id-1",
              "name": "rootTask-for-project-name-1",
            },
          },
        },
        Object {
          "cursor": "project-id-2",
          "node": Object {
            "name": "project-name-2",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-2",
                "name": "project-name-2",
              },
              "id": "rootTask-for-project-id-2",
              "name": "rootTask-for-project-name-2",
            },
          },
        },
        Object {
          "cursor": "project-id-3",
          "node": Object {
            "name": "project-name-3",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-3",
                "name": "project-name-3",
              },
              "id": "rootTask-for-project-id-3",
              "name": "rootTask-for-project-name-3",
            },
          },
        },
        Object {
          "cursor": "project-id-4",
          "node": Object {
            "name": "project-name-4",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-4",
                "name": "project-name-4",
              },
              "id": "rootTask-for-project-id-4",
              "name": "rootTask-for-project-name-4",
            },
          },
        },
        Object {
          "cursor": "project-id-5",
          "node": Object {
            "name": "project-name-5",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-5",
                "name": "project-name-5",
              },
              "id": "rootTask-for-project-id-5",
              "name": "rootTask-for-project-name-5",
            },
          },
        },
        Object {
          "cursor": "project-id-6",
          "node": Object {
            "name": "project-name-6",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-6",
                "name": "project-name-6",
              },
              "id": "rootTask-for-project-id-6",
              "name": "rootTask-for-project-name-6",
            },
          },
        },
        Object {
          "cursor": "project-id-7",
          "node": Object {
            "name": "project-name-7",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-7",
                "name": "project-name-7",
              },
              "id": "rootTask-for-project-id-7",
              "name": "rootTask-for-project-name-7",
            },
          },
        },
        Object {
          "cursor": "project-id-8",
          "node": Object {
            "name": "project-name-8",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-8",
                "name": "project-name-8",
              },
              "id": "rootTask-for-project-id-8",
              "name": "rootTask-for-project-name-8",
            },
          },
        },
        Object {
          "cursor": "project-id-9",
          "node": Object {
            "name": "project-name-9",
            "rootTask": Object {
              "containingProject": Object {
                "id": "project-id-9",
                "name": "project-name-9",
              },
              "id": "rootTask-for-project-id-9",
              "name": "rootTask-for-project-name-9",
            },
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
      },
    },
  },
}
`;

exports[`run for GetTasksInProjectDocument 1`] = `
Object {
  "defaultDocument": Object {
    "projects": Object {
      "byId": Object {
        "rootTask": Object {
          "tasks": Object {
            "edges": Array [
              Object {
                "node": Object {
                  "completed": false,
                  "containingProject": Object {
                    "id": "some-project-id",
                    "name": "projectName",
                  },
                  "effectiveDueDate": null,
                  "effectivelyCompleted": false,
                  "flagged": false,
                  "id": "bar",
                  "name": "foo",
                },
              },
            ],
          },
        },
      },
    },
  },
}
`;
